src/plugins/unified_histogram/public/services/lens_vis_service.ts

```diff
@@ -0,0 +1,701 @@
+      allSuggestions: undefined,
+      currentSuggestionContext: {
+        suggestion: undefined,
+        suggestionDeps: undefined,
+        type: UnifiedHistogramSuggestionType.unsupported,
+      },
+      lensAttributesContext: undefined,
+    });
+
+    const stateSelector = stateSelectorFactory(this.state$);
+    this.status$ = stateSelector((state) => state.status);
+    this.allSuggestions$ = stateSelector((state) => state.allSuggestions);
+    this.currentSuggestionContext$ = stateSelector(
+      (state) => state.currentSuggestionContext,
+      isEqual
+    );
+    this.lensAttributesContext$ = stateSelector((state) => state.lensAttributesContext, isEqual);
+    this.prevUpdateContext = undefined;
+  }
+
+  update = ({
+    externalVisContext,
+    queryParams,
+    chartTitle,
+    timeInterval,
+    breakdownField,
+    onSuggestionContextChange,
+    onVisContextChanged,
+  }: {
+    externalVisContext: ExternalVisContext | undefined;
+    queryParams: QueryParams;
+    chartTitle: string | undefined;
+    timeInterval: string | undefined;
+    breakdownField: DataViewField | undefined;
+    onSuggestionContextChange?: (
+      suggestionContext: UnifiedHistogramSuggestionContext | undefined
+    ) => void;
+    onVisContextChanged?: (visContext: ExternalVisContext | undefined) => void;
+  }) => {
+    const suggestionContextSelectedPreviously = this.state$.getValue().currentSuggestionContext;
+
+    // console.log('recalculating chart', queryParams.query, externalVisContext);
+
+    const allSuggestions = this.getAllSuggestions({ queryParams });
+
+    // console.log('service allSuggestions', allSuggestions);
+
+    const suggestionState = this.getCurrentSuggestionState({
+      suggestionContextSelectedPreviously,
+      externalVisContext,
+      queryParams,
+      allSuggestions,
+      timeInterval,
+      breakdownField,
+    });
+
+    // console.log('service suggestionState', suggestionState);
+
+    const lensAttributesState = this.getLensAttributesState({
+      currentSuggestionContext: suggestionState.currentSuggestionContext,
+      externalVisContext,
+      queryParams,
+      chartTitle,
+      timeInterval,
+      breakdownField,
+    });
+
+    // console.log('service lensAttributesState', lensAttributesState);
+
+    if (suggestionState.shouldUpdateSelectedSuggestionDueToDepsChange) {
+      onSuggestionContextChange?.(suggestionState.currentSuggestionContext);
+    }
+
+    if (
+      externalVisContext?.attributes &&
+      (suggestionState.shouldUpdateSelectedSuggestionDueToDepsChange ||
+        lensAttributesState.shouldUpdateVisContextDueToIncompatibleSuggestion)
+    ) {
+      // console.log(
+      //   'forced to update selected suggestion and vis context',
+      //   suggestionState.currentSuggestionContext.suggestion,
+      //   lensAttributesState.lensAttributesContext
+      // );
+      onVisContextChanged?.(lensAttributesState.lensAttributesContext);
+    }
```

Comment:
Invalidation logic to reset suggestion and saved vis context and they are not compatible any more to the query context.